### PR TITLE
Migrate source files to TypeScript

### LINKED ISSUE #20: Migrate to TypeScript
Migrate the project to TypeScript. This will allow generating type definitions files instead of writing them by hand.

### PR DESCRIPTION
**Pull request by whymarrh**
Closes #20

This change migrates the source files to TypeScript, leaving just the tests written in JS. This allows us to auto-generate type definitions for the project and to ensure a basic set of type safety.

The test files can be updated in a separate PR, optionally.

### CODE REVIEW

File: lib/composed.ts [hunk 1]
```diff
@@ -1,10 +1,12 @@
 'use strict'
 
-const ObservableStore = require('.')
+import ObservableStore = require('.')
```
Comment (rekmarks) [inline lib/composed.ts:3]: What's this syntax? 👀

File: package.json [hunk 2]
```diff
@@ -3,27 +3,33 @@
   "version": "4.0.3",
   "description": "`ObservableStore` is a synchronous in-memory store for a single value, that you can subscribe to updates on.",
   "main": "lib/index.js",
+  "types": "lib/index.d.ts",
   "files": [
-    "lib/*.js"
+    "lib/*.js",
+    "llib/*.d.ts"
```
Comment (rekmarks) [inline package.json:9]: Typo:
```suggestion
    "lib/*.d.ts"
```

File: lib/composed.ts (hunk 1 above)
Comment (whymarrh) [reply lib/composed.ts:3]: [This is the import for the corresponding `export =` syntax.](https://www.typescriptlang.org/docs/handbook/modules.html#export--and-import--require)

File: lib/localStorage.ts [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
+import ObservableStore = require('.')
 
 class LocalStorageStore extends ObservableStore {
 
-  constructor (opts = {}) {
-    if (!global.localStorage) {
+  private _storageKey: string;
+
+  constructor (opts: { storageKey?: string } = {}) {
+    if (!(global as any).localStorage) {
```
Comment (rekmarks) [inline lib/localStorage.ts:10]: Should we do this here as well?
```suggestion
    if (!(global as unknown as Window).localStorage) {
```

File: lib/asStream.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  }
+
+  // emit current state on new destination
+  pipe<T extends NodeJS.WritableStream>(dest: T, options?: { end?: boolean }): T {
+    let result = super.pipe(dest, options);
+    dest.write(this.obsStore.getState() as any)
+    return result;
+  }
+
+  // write from incomming stream to state
```
Comment (rekmarks) [inline lib/asStream.ts:32]: This typo has irked me in the past!
```suggestion
  // write from incoming stream to state
```

File: .eslintrc.js [hunk 5]
```diff
@@ -7,6 +7,10 @@ module.exports = {
     '@metamask/eslint-config',
     '@metamask/eslint-config/config/nodejs',
   ],
+  rules: {
+    // TODO: Re-enable this post-TS migration
+    'import/no-unresolved': 'off',
+  },
```
Comment (rekmarks) [inline .eslintrc.js:13]: Just for my own understanding, why do we have to do this?

**Review by rekmarks — CHANGES_REQUESTED** [review]
Neat!

Just some questions and nits.

File: lib/composed.ts (hunk 1 above)
Comment (rekmarks) [reply lib/composed.ts:3]: 💡

File: .eslintrc.js (hunk 5 above)
Comment (whymarrh) [reply .eslintrc.js:13]: ESLint should be able to run without the built files 🤔 but this comment doesn't make any sense

Comment (whymarrh) [reply .eslintrc.js:13]: I'll try and remove this, and add the a the prelint step run the build

File: package.json [hunk 6]
```diff
@@ -3,27 +3,33 @@
...
-    "lib/*.js"
+    "lib/*.js",
+    "lib/*.d.ts"
   ],
   "scripts": {
-    "test": "tape test/**/*.js",
-    "lint": "eslint .",
+    "build": "tsc --project .",
+    "test": "yarn build && tape test/**/*.js",
+    "lint": "yarn build && eslint test/*.js",
```
Comment (Gudahtt) [inline package.json:14]: It looks like the `.ts` files aren't being linted? :thinking:

Comment (whymarrh) [reply package.json:14]: They aren't, no. That's a separate PR.

Comment (whymarrh) [reply package.json:14]: I wanted to make minimal changes to the JS source to facilitate review. Linting the TS sources would add a lot of changes/noise to the diff.

**Review by rekmarks — APPROVED** [review]
LGTM!

**Pull request merged**